[리액트 7기 표후동] React routing

ndopy opened the pull request:
안녕하세요.
4주차~5주차를 여러 번 복습해서 이해하느라 과제가 늦어졌네요.

아직도 부족하지만 매일 구현한만큼이라도 피드백 요청드릴수 있도록 노력해보겠습니다.

Commit:
config: devServer.historyApiFallback 설정을 추가한다.

- 존재하지 않는 페이지에 접근할 때 index.html 로 redirect 한다.

Commit:
test: <RestaurantsPage /> 에 대한 테스트 작성

- <App/> 의 컴포넌트들을 옮길 <RestaurantsPage /> 에 대한 테스트를 작성한다.

Commit:
add: <RestaurantsPage /> 작성

- <App/> 의 컴포넌트들을 옮긴다.

Commit:
refactor: <App />

- 기존 컴포넌트들을 <RestaurantsPage /> 컴포넌트로 대체한다.

Commit:
refactor: <App />

- 사용하지 않는 코드들을 제거한다.

Commit:
refactor: <App/> test

- 불필요한 테스트 코드를 제거한다.

Commit:
feat: 간단한 router 구현해보기

- <a> 태그를 이용해서 화면을 이동하는 router 를 작성한다.

Commit:
refactor: <App/>에서 HomePage 를 분리한다.

- <HomePage /> 를 별도의 파일로 분리한다.

Commit:
test: <HomePage/> 를 렌더링할 수 있다.

- <HomePage/>가 렌더링되어 'About' 이라는 텍스트가 있는지 확인한다.

Commit:
test: <AboutPage/> 에 대한 테스트 작성

- <AboutPage /> 를 렌더링할 수 있다.

Commit:
refactor: <App /> 에서 <AboutPage/> 를 분리한다.

Commit:
test: <NotFoundPage/> 테스트 코드 작성

- <NotFoundPage/> 를 렌더링할 수 있다.

Commit:
refactor: <App/> 에서 <NotFoundPage/> 를 분리한다.

Commit:
build: install react-router-dom

- npm install react-router-dom 명령어를 실행한다.

Commit:
feat: react-router-dom 을 이용한 라우팅 구현

- <BrowserRouter>, <Routes>, <Route> 를 이용해 라우팅을 구현한다.

-  V6 이전에는 <Routes> 대신 <Switch> 를 사용했다.
  ㄴ 'Upgrading from v5' -> 'Upgrade all <Switch> elements to <Routes>' 부분을 읽어보자.
  ㄴ https://reactrouterdotcom.fly.dev/docs/en/v6/upgrading/v5#upgrade-all-switch-elements-to-routes

Commit:
refactor: <HomePage/>의 <a>를 <Link> 로 변경한다.

- react-router-dom 에서는 페이지를 네비게이션할 때
  <Link to="주소"></Link> 를 사용한다.

Commit:
fix: <HomePage/> 테스트를 수정한다.

- <Link> 를 사용하려면 <BrowserRouter> 안에서 사용해야 한다.
  -> 따라서 테스트 코드의 <HomePage/> 를 <BrowserRouter>로 감싸준다.

Commit:
refactor: <HomePage/>의 <BrowserRouter>를 <MemoryRouter>로 변경한다.

- <MemoryRouter>
 ㄴ Router 와 같은 역할을 하지만,
    주소 표시줄에 url 을 나타내지 않는 대신 메모리에 기록한다고 한다.

Commit:
refactor: <BrowserRouter> 를 index.jsx 로 옮긴다.

- <App/> 의 <BrowserRouter> 를 index.jsx 로 옮긴다.
- <App/> 테스트에서도 <App/> 을 <MemoryRouter> 로 감싸줘야 한다.

Commit:
refactor: <App/> 테스트 코드를 리팩토링한다.

- 공통으로 실행할 부분을 beforeEach 로 분리한다.
- 테스트별 상황을 context 로 분리한다.

Commit:
test: /about 에 대한 테스트를 작성한다.

- MemoryRouter 의 initialEntries 에 경로를 등록할 수 있다.

Commit:
test: <App/> 테스트를 리팩토링한다.

- <MemoryRouter> 를 render 하는 코드가 반복적으로 사용된다.
  ㄴ renderApp 함수로 분리한다.

Commit:
test: /restaurants 라우팅에 대한 테스트를 작성한다.

- 해당 경로로 이동했을 때 '서울' 이라는 텍스트가 보이는지 확인한다.
- <Restaurants/> 에서 dispatch 를 사용하므로
  테스트 코드에서도 useDispatch 를 구현해야 한다.
- regions 상태 안에 값이 있어야 확인할 수 있으므로,
  테스트에서 확인할 값을 regions 상태에 추가해야 한다.

Commit:
test: 기타 다른 경로(ex. /xxx) 에 대한 테스트 코드를 작성한다.

- /xxx 같은 경로로 이동 시, '404'라는 텍스트가 있는지 확인한다.

Commit:
add: <App/> 의 Route 에 기타 경로(/*)를 추가한다.

- 미리 선언되지 않은 경로로 접근 시, NotFoundPage 를 보여준다.
  ㄴ /, /about, /restaurants 경로는 미리 선언되어 있지만
    /xxx 와 같은 경로는 미리 선언되지 않았기 때문에
    NotFoundPage 로 이동한다.

Commit:
Add failing test : RestaurantDetailPage

Commit:
Pass failing test : RestaurantDetailPage
- need to improve the code

Commit:
Refactor filename: HoomPage to HomePage

Commit:
Add restaurantInformation into initialState in reducer

- Add restaurantInformation into reducer.test.jsx too.

Commit:
Add restaurantInformation fixture

- make restaurantInformation.js in fixtures

moonkii commented on src/App.test.jsx:
```diff
@@ -1,34 +1,63 @@
+    });
+  });
+
+  context('with other path', () => {
```
other path 라는건 어떤 의미인가요?

moonkii reviewed (commented):
![image](https://user-images.githubusercontent.com/46776329/165760588-ca28c2aa-18c7-44bf-b930-5932806e5106.png)

잘 진행하고 있으신 것 같아요!
나머지 부분도 마저 완성해서 올려주세요 👍

Commit:
Style : App.test.jsx 의 포맷팅 변경

Commit:
Refactor : 테스트 코드 리팩토링

- beforeEach 로 dispatch mock 을 매 테스트마다 초기화한다.

Commit:
Test: RestaurantsPage renders links of restaurants

Commit:
Fix: toHaveTextContent -> toContain

Commit:
Passing failing test : RestaurantsPage renders links of restaurants

Commit:
Refactor: <MemoryRouter> 를 사용한다.

- <Link> 를 사용하려면 <Link> 를 사용하려는 컴포넌트와 상위 컴포넌트를 <BrowserRouter> 로 감싸줘야 한다.
- 테스트 코드에서는 <BrowserRouter> 대신 <MemoryRouter> 를 사용한다.

Commit:
Refactor: <a> 대신 <Link> 를 사용한다.

Commit:
Refactor: renderRestaurantPage 로 분리한다.

Commit:
Refactor: RestaurantsContainer.test

- <RestaurantsContainer/> 에 <Link> 자체를 넘기도록 한다.

- 아래 코드의 내용은 잘 이해가 안된다.
  - const Link = ({ children }) => (<>{children}</>);

Commit:
Add: RestaurantsPage 에서 RestaurantsContainer 로 Link 를 전달하도록 구현한다.

- RestaurantsPage : Link 를 RestaurantsContainer 로 전달한다.
- RestaurantsContainer : RestaurantsPage 에서 전달받은 Link 를 <Link to=""> 로 사용한다.

Commit:
Refactor: RestaurantsContainer test

- 아샬님이 가장 선호하는 방식
-> "움직이는 것을 직접 내려주는 방식"
-> "뭔가를 누르면 어떻게 움직인다" 라는 동작을 내려주는 것

- 이렇게 이해하는 것이 맞을까?
-> onClickRestaurant={handleClick}
  ㄴ 뭔가(=Restaurant)를 누르면(onClick)
  ㄴ 어떻게 움직인다(=handleClick)

ndopy commented on src/RestaurantsContainer.test.jsx:
```diff
@@ -13,9 +13,11 @@ test('RestaurantsContainer', () => {
-      <RestaurantsContainer />
+      <RestaurantsContainer Link={Link} />
     </MemoryRouter>
   ));
```
과제 풀이 영상을 보니까
아샬님이 이러한 코드로 Link 를 전달하는 내용을 구현하셨습니다.

`<Link>` 를 `({ children }) => (<>{children}</>)` 로 작성하신 것이 잘 이해가 되지 않는데
실제 `<Link>` 가 저런 방식으로 동작해서인가요?

위 내용을 이해하려면 어디서부터 접근하면 좋을까요?

moonkii replied:
`children` 와 `Link` 와는 별개에요. 
`children` 이 무엇인지 부터 찾아보시면 좋을 것 같아요.
공식문서에 보시면 `children` 이 나올거에요.

Commit:
Add: RestaurantContainer 가 전달 받을 onClickRestaurant 을 어떻게 사용할 것인지에 대한 내용 추가

- <a> 로 작성하면 기본적으로 화면이 아예 이동해버린다.
 ㄴ event.preventDefault() 로 해당 기본 동작을 막는다.
 ㄴ 이제 레스토랑 이름을 클릭해도 화면 이동이 되지 않는다.

Commit:
fix: event.preventDefault() 가 적용되지 않고 있다.

- onClick={() => handleClick(restaurant)}
 -> onClick={handleClick(restaurant)} 으로 수정한다.

ndopy commented on src/App.test.jsx:
```diff
@@ -1,34 +1,66 @@
-
 import { render } from '@testing-library/react';
 
 import { useDispatch, useSelector } from 'react-redux';
 
+import { MemoryRouter } from 'react-router-dom';
 import App from './App';
 
-test('App', () => {
-  const dispatch = jest.fn();
+describe('App', () => {
+  beforeEach(() => {
+    const dispatch = jest.fn();
+
+    useDispatch.mockImplementation(() => dispatch);
+
+    useSelector.mockImplementation((selector) => selector({
+      regions: [
+        {
+          id: 1,
+          name: '서울',
+        },
+      ],
+      categories: [],
+      restaurants: [],
+    }));
+  });
```
테스트 코드 내에 jest.mock('react-redux') 가 작성되지 않았는데도
~.mockImplementation 이 동작할 수 있는 이유가 궁금합니다.

이전 강의들을 통해 제가 이해한 내용은
" jest.mock('react-redux') 를 작성해야
jest 에서 'react-redux.js' 라는 파일에 작성된 `jest.fn()` 들을 mocking 하여
`~.implementation` 을 통해 구현할 수 있다" 입니다.

그런데 여기에는 `jest.mock('react-redux')` 을 선언한 코드가 없는데
`useDispatch` 와 `useSelector` 의 `mockImpelementation` 이 동작하는 게
이상하다고 생각이 들어서 질문드립니다.

moonkii replied:
`react-redux` 가 `__mocks__` 에 들어가있으면 아마 jest 에서 자동으로 인식해줘서 그럴거에요

Commit:
Add: /restaurants path 를 exact 하게 설정한다.

Commit:
Add: useNavigate 로 페이지 이동 구현

- react-router-dom V6부터 useHistory -> useNavigate 로 대체되었다.

Commit:
Failing test: RestaurantDetailPage renders name of restaurant

- 강의 풀의 내용대로 따라서 구현해보자.

Commit:
Refactor test : RestaurantDetailPage renders name of restaurant

[과제 풀이 영상 설명]
- 주소를 가지고 나누는 방식은 <App/> 에서 담당하는 것이다.
- <RestaurantDetailPage /> 는 <App />에서부터 Router 를 타고 들어온 게 아니다.
- 따라서 "restaurants/1"은 별도로 처리해줘야 한다.

Commit:
Pass failing test : RestaurantDetailPage
- useParams : React Router 에서 사용하는 파라미터들을 확인할 수 있는 Hook 이다.
- RestaurantDetailPage 에 전달받은 params 가 있다면 해당 params 를 사용하고, 그렇지 않다면 useParams 를 사용하도록 구현한다.
- RestaurantDetailPage 에 전달할 params 는 "restaurantId" 이다.

Commit:
Add: fetchRestaurant into api

- restaurantId 를 기준으로 특정 레스토랑의 상세 정보를 얻기 위한 API 요청을 추가한다.

Commit:
Refactor: api.js

- API 요청 시 사용하는 공통된 주소를 BASE_URL 로 분리한다.

Commit:
Add: loadRestaurant

- RestaurantDetailPage 가 렌더링될 때 특정 레스토랑 정보를 불러오는 loadRestaurant 함수를 추가한다.
- 이 다음에는 dispatch 를 통해서 액션을 구현해야 한다.

Commit:
Add: loadRestaurant, setRestaurant Action 추가

Commit:
Add: RestaurantContainer, RestaurantDetail

- RestaurantContainer : React Router 를 몰라도 된다.
- RestaurantDetail : Redux 를 몰라도 된다. RestaurantContainer 로부터 전달받은 props 를 사용할 뿐이다.
- 분리하기 전에 function 으로 구현한다.
  ㄴ 이후에 별도 컴포넌트로 분리할 것.
- useSelector 를 이용해서 Redux store 에서 restaurant 상태를 가져온다.

Commit:
Refactor: 파일명 변경

- RestaurantDetailPage -> RestaurantPage

Commit:
Refactor: restaurantInformation -> restaurant

- 레스토랑 상세 정보를 저장할 state 이름을 변경한다.

Commit:
Add: useSelector, useDispatch mockImplementation in RestaurantPage test

Commit:
Fix: RestaurantPage

- 레스토랑 상세 정보 state 를 restaurant 로 정했다.
 ㄴ RestaurantDetail 컴포넌트에 전달할 데이터도 레스토랑의 상세 정보이므로 "restaurantId" 가 아니라 "restaurant" 여야 한다.

 - RestaurantDetail 에서는 레스토랑 상세 정보를 보여줘야 한다.
  ㄴ 레스토랑의 id 는 "restaurantId" 라고 정했다.
  ㄴ 따라서 상세 정보 화면에서 보여줄 레스토랑의 id 도  "restaurant.Id" 가 아니라 "restaurant.restaurantId "가 되어야 한다.

Commit:
Fix: fetchRestaurant in api

- fetchRestaurant 는 특정 restaurantId 를 전달받고, 해당 restaurantId 의 레스토랑 상세 정보를 조회하는 동작을 수행한다.
- 레스토랑의 ID를 "restaurantId" 로 사용하기로 했기 때문에 "id" -> "restaurantId" 로 수정한다.

hdpyo commented:
구현한 내용 중에 궁금한 내용이 있어서 위에 셀프리뷰로 작성해두었습니다.

오늘(4/30 토)부터 과제 풀이 영상을 보면서 아샬님 따라서 구현해보기 시작했는데,
이전 주차의 과제들도 "우선 완벽히 이해가 안되더라도 구현해가며 이해하려고 해보자!" 라고 했다보니,
풀이 영상을 보면서도 헷갈리거나 어려운 점이 있었습니다.

저에게 필요한 건 역시 반복숙달인 것 같습니다.
Redux, 비동기, 라우팅 등 이전 주차의 강의를 계속 복습하고,
아샬님의 가르쳐주시는 좋은 코딩 패턴을 체화할 수 있도록 반복 작업해야할 것 같습니다.

moonkii commented:
> 저에게 필요한 건 역시 반복숙달인 것 같습니다.
Redux, 비동기, 라우팅 등 이전 주차의 강의를 계속 복습하고,
아샬님의 가르쳐주시는 좋은 코딩 패턴을 체화할 수 있도록 반복 작업해야할 것 같습니다.

정확하게 짚으셨어요!
지금 당장 아샬님의 강의와 내포된 메시지를 모두 이해하는게 정말 쉽지 않아요.
일단은 최소한 코드 패턴이라도 외워서 구현이 가능하도록 하신 뒤에 점점 숙련도가 쌓이고 실력이 늘어가면서
자연스럽게 내용이나 메시지가 이해되실거에요.
저 조차도 수십번을 반복해서 보아왔지만 아직도 배움을 얻어가는 중이에요!
